sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShimLoader.scala

```diff
@@ -119,41 +125,99 @@ object ShimLoader extends Logging {
     s"org.apache.spark.sql.rapids.shims.$shimId.RapidsShuffleInternalManager"
   }
 
+  private def serializerClassloader() = {
+    // Hypothesis: serializer is the most universal way to intercept classloaders
+
+    // https://github.com/apache/spark/blob/master/core/src/main/scala/
+    // org/apache/spark/serializer/JavaSerializer.scala#L147
+
+    // https://github.com/apache/spark/blob/master/core/src/main/scala/
+    // org/apache/spark/serializer/KryoSerializer.scala#L134
+
+    Option(SparkEnv.get)
+      .map(_.serializer)
+      .flatMap { serializer =>
+        logInfo("Looking for a mutable classloader (defaultClassLoader) in SparkEnv.serializer " +
+          serializer)
+        // scalac generates accessor methods
+        val serdeClassLoader = MethodUtils
+          .invokeMethod(serializer, true, "defaultClassLoader")
+          .asInstanceOf[Option[ClassLoader]]
+          .getOrElse {
+            val threadContextClassLoader = Thread.currentThread().getContextClassLoader
+            logInfo(s"No defaultClassLoader found in $serializer, falling back " +
+              s"on Thread context classloader: " + threadContextClassLoader)
+            threadContextClassLoader
+          }
+
+        logInfo("Extracted Spark classloader from SparkEnv.serializer " + serdeClassLoader)
+        findURLClassLoader(serdeClassLoader)
+      }.orElse {
+        logInfo("Spark-less use case: RapidsConf.help?")
+        Option(getClass.getClassLoader)
+      }
+  }
+
+
+  @tailrec
+  private def findURLClassLoader(classLoader: ClassLoader): Option[ClassLoader] = {
+    // walk up the classloader hierarchy until we hit a classloader we can mutate
+    // in the upstream Spark, non-REPL/batch mode serdeClassLoader is already mutable
+    // in REPL use-cases, and blackbox Spark apps it may take several iterations
+
+    // ignore different flavors of URL classloaders in different REPLs
+    // brute-force call addURL using reflection
+    classLoader match {
+      case nullClassLoader if nullClassLoader == null =>
+        logInfo("findURLClassLoader failed to locate a mutable classloader")
+        None
+      case urlCl: java.net.URLClassLoader =>
+        // fast path
+        logInfo(s"findURLClassLoader found a URLClassLoader $urlCl")
+        Option(urlCl)
+      case replCl if replCl.getClass.getName == "org.apache.spark.repl.ExecutorClassLoader" =>
+        // https://issues.apache.org/jira/browse/SPARK-18646
+        val parentLoader = MethodUtils.invokeMethod(replCl, true, "parentLoader")
+          .asInstanceOf[ClassLoader]
+        logInfo(s"findURLClassLoader found $replCl, trying parentLoader=$parentLoader")
+        findURLClassLoader(parentLoader)
+      case urlAddable: ClassLoader if null != MethodUtils.getMatchingMethod(
```

Comment:
should: `case urlCl: java.net.URLClassLoader` not match the class loader you are matching in the case at this line (184)?

Reply:
And if both are instances of `URLClassLoader` should the method return `Option[URLClassLoader]`?